### ISSUE TITLE
Feature: Auth completed callback on SSHServer

### ISSUE THREAD

**Issue by matthijskooijman** [ronf/asyncssh#152]
I'm implementing an SSH server, where I'm using key auth. After auth completes, I want to process some custom options from the `authorized_keys` file and do some one-time calculations based on that.

I noticed `SSHServer.set_authorized_keys()`, which conveniently lets `SSHServer` handle all the steps of reading the file and checking each key. Afterwards, I can use `SSHServer.get_key_option()` to access the options of the key used for authentication. However, as it is now, my code does not get notified when the authentication step completes, so I have no good place to do process the options. I see two possible implementations: 1) process each key auth manually, so I can process before returning succesful authentication, or 2) access the result of my key option processing through various methods, which do the processing if it was not done already. Neither of these options are really elegant.

A simple way to fix this, would be to end a `complete_auth()` callback on `SSHServer` (which would match with `begin_auth()`). Note that this call always means that authentication is successful, since the server never decides to completely fail authentication (typically, the client decides to stop trying after it exhausted all options).

Looking at the code, `SSHServerConnection.send_userauth_success()` seems an obvious place to put this call, though I'm not sure if that is the best place?

**Comment by ronf** [comment]
It's not at the connection level, but one thing that may work for you here is to do this processing either in the session_started() callback in a custom SSHServerSession class or to do it at the very beginning of the method you pass in as the session_factory or process_factory. One benefit to doing this is that you'll have a way to provide feedback to the user if you end up deciding to close the connection in some cases. If you are supporting multiple sessions on a connection and want to do this processing only once, it's easy enough to set a flag indicating that it was already done.

On the client side, there is an auth_completed() method which is called whenever the authentication has completed successfully. It was intended as the place to trigger opening of sessions, but it is largely unnecessary now as the await that is typically done to open the SSH connection will block now until just after auth_completed() is called, so it's cleaner to simply put this code right after the call to create_connection() or equivalent. This callback dates back to the asyncore version of AsyncSSH where everything had to be done with callbacks and exists now only to maintain backward compatibility. I could mirror this on the server side, but I'm not sure how useful it would be. The only benefit would be that it could run without waiting for the client to request its first session, but there'd be no way to provide feedback to the client at that point.

In your use case, is it important to run this even before the client requests its first session? If so, are you ok not being able to provide feedback?

I guess one possible issue here is if your clients are not requesting sessions and are instead requesting things like port forwarding. However, you could put a call to your additional processing in the server's connection_requested() or server_requested() messages in this case.

**Comment by matthijskooijman** [comment]
>  to do it at the very beginning of the method you pass in as the session_factory or process_factory

That would be a variation of option 2). Also, I'm heavily using port forwards as well, which do not exist within a session, so I'd end up calling the "process-if-needed" function in a handful of places (which works, but is a bit fragile and not very elegant).

> In your use case, is it important to run this even before the client requests its first session? If so, are you ok not being able to provide feedback?

It do not really require running this early, though it would be convenient to implicitly run the processing exactly once (e.g without having to set a flag, and without allowing me to forget the processing). What I need to do is process some key options and set some member variables based on the option, to be used later.

Providing feedback is not relevant now, though now you mention it, I might need that later (but in that case I would have to process key auth manually anyway, if I want to deny keys based on options. And I'll likely  not actually need that, not sure yet).

So: Adding an `auth_completed()` method to mirror the client does look useful to me (I'd use it, even if it does not have any way to provide feedback), though doing key auth manually is not a big deal either. I'll leave it up to you to decide if this is a worthwhile addition to the API or not :-)

**Comment by ronf** [comment]
I've just checked in 2e0a009 in the "develop" branch which should provide this functionality. If you get a chance to try it out, let me know if this does what you were looking for.

**Comment by ronf** [comment]
I had a fix for another issue I wanted to get out in a patch release, so I've gone ahead and rolled this change into the 1.13.2 release. If you run into any problems with it, let me know!
